### DISCUSSION TITLE
Compiler's suggestions for better performance

### DISCUSSION THREAD

**Discussion by gryznar** [modular/modular#323] [Ideas]
There is a a group of people which like squeeze as much performance from their code as possible. But these people may come directly from Python, without any technical background beyond about low level things or which just do not know about fact, that some cool dynamic features just kills performance or which simple are not aware about Mojo's possiblities in optimization.

I wish that Mojo's compiler could some day be smart enough to give tips and advices what to do to have better performance  (with special flag or in special mode), e.g. why to avoid some things, what to do, etc.

I would really appreciaty this to become myself  better and more concious programmer.

**Comment by lattner** [comment]
Mojo already gives a couple warnings that suggest better things to do (e.g. use let instead of var where possible).  that said, the compiler isn't good at pointing out larger design pattern changes, for this I think we'll have LLM based tools outside the compiler itself.  The UI is much better for explaining things in that context

**Comment by gryznar** [comment]
Great to hear that Modular aims to deliver such tools! I really, really, really appreciate that! 

So simple things in compiler, more advanced in external tools? Sounds reasonable ;)  Of course, my first thoughts were about simple things to explain and to warn about in compiler, not the code refactoring. Things which would work, because of Python compatibility, but which are not so efficient (redeclaring type of variable with the same name for example)
```python
a = 1
a = '1' 
```
I think  that Mojo's compiler (no matter how well it will be optimized) will still suffer from staff like that. I mean such simple, but not obvious cases for programmers who get used to dynamic nature of Python.
